### ISSUE TITLE
Analysis/Ingestion taking a very long time for some packages

### ISSUE THREAD

**Issue by daveverwer** [SwiftPackageIndex/SwiftPackageIndex-Server#3290]
I pushed a change to Leftpad's main branch at around midday yesterday and as of 6pm today, 30 hours later, the main branch change still has not been picked up by analysis and the README change still has not been picked up by ingestion.

![Screenshot 2024-08-13 at 18 41 45@2x](https://github.com/user-attachments/assets/fdab3992-dc7c-4ab7-8d82-fa60cac733da)

On staging, both changes were picked up in reasonable time.

**Comment by daveverwer** [comment]
I thought it might be something to do with #3285 at first, but it now looks unrelated.

**Comment by daveverwer** [comment]
In `packages` on prod:

![Screenshot 2024-08-13 at 18 47 34@2x](https://github.com/user-attachments/assets/2c340ee8-41a2-45a5-b118-493b71f4ad98)

**Comment by finestructure** [comment]
Generally, we analyse packages every ~6h at the moment:

![CleanShot 2024-08-15 at 11 04 04@2x](https://github.com/user-attachments/assets/7ecfdd0b-5039-4027-a808-7cc296b9515d)

For some reason, LeftPad isn't being picked up for analysis:

![CleanShot 2024-08-15 at 11 19 41@2x](https://github.com/user-attachments/assets/689e22c7-251c-4f55-bf1c-81ddf539718e)

It _does_ show up in the candidate selection query and I saw its position move up. I was around 515 at 10:30 CET this morning and now it's back down the bottom > position 5000. Its `packages.update_at` has moved to `2024-08-15 08:37:48.362663+00`, so it _has_ been visited by the process. `processing_stage` is `ingestion` which it would be after analysis has processed it.

**Comment by finestructure** [comment]
I don't think it can be ingestion, because ingestion doesn't pick up packages that are in `ingestion` stage. There's no indication that it moved out of `ingestion`. If analysis had picked it up, we'd a) see the new version, b) see its line in the logs, and c) see the stage at `analysis`.

Reconciliation doesn't look at existing packages at all, it's entirely driven by the package list as input.

I can't think of another process that's updating the `packages` table off the top of my heads but there must be something touching it.

**Comment by finestructure** [comment]
I've reviewed where we're updating the `packages` table and unless I'm missing something there's only ingestion and analysis.

**Comment by finestructure** [comment]
I think I have a lead. Checking what are the latest updated records across the whole scheme for LeftPad, I noticed that there are build records at exactly the same time the package was updated `2024-08-15 08:37:48.332068+00`:

![CleanShot 2024-08-15 at 12 34 47@2x](https://github.com/user-attachments/assets/73fc85e1-9eda-426c-b1ac-cbbbd819bbca)

Notice the infrastructure error and that these are visionOS builds. I suspect we have two issues:

- Build reports not only update the builds table but also touch `packages.updated_at`. That'd be surprising but it certainly looks like it. Or it only happens in error cases? Needs investigation.
- The visionOS infrastructure error is problematic, because it'll recur every 4 hours or so when we retry them.

This would explain what's going on: The retried visionOS builds update the package and put it back at the end of the analysis queue before it gets a chance to be analysed.

**Comment by finestructure** [comment]
This would also explain why this is only happening now. At least I'm pretty sure we'd have seen reports for outdated packages if this had been going on for longer. It's likely related to our latest Swift 6 run which exposes the package updating issue on build reports.

This build report updating issue (while not ideal) is not a problem in general, because even though it'll bump a package back to the end of the queue this will only happen once when the build is done. Here though, because the build is being retried continuously on a 4h schedule, it happens over and over before analysis can kick in.

**Comment by finestructure** [comment]
Tracking infrastructure errors in different in new issue: #3294

**Comment by finestructure** [comment]
We are indeed updating the package when we're [receiving a build report](https://github.com/SwiftPackageIndex/SwiftPackageIndex-Server/blob/4556da48bfe9f431ca29caa8079081a0e505355e/Sources/App/Controllers/API/API%2BBuildController.swift#L90):

```
                try await Package.updatePlatformCompatibility(for: version.$package.id, on: req.db)
```

Will have to see how/if we can avoid the re-queueing from happening.

**Comment by daveverwer** [comment]
> Will have to see how/if we can avoid the re-queueing from happening.

It might be time for explicit `last_analysed_at` and `last_ingested_at` fields.
